src/python/pants/backend/python/subsystems/pex_bootstrap/_ptex_launcher.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+"""Entrypoint script for a .ptex file generated with --generate-ptex.
+
+This script will build a .ipex file into a temporary directory, then execute it. If
+PANTS_PTEX_HYDRATE_ONLY_TO_IPEX=<filename> is set, however, the script will generate the ipex at
+<filename>, then exit without executing the .ipex file.
+"""
+
+import json
+import os
+import sys
+import tempfile
+
+from pex import resolver
+from pex.common import open_zip
+from pex.fetcher import Fetcher, PyPIFetcher
+from pex.interpreter import PythonInterpreter
+from pex.pex_builder import PEXBuilder
+from pex.pex_info import PexInfo
+
+
+self = sys.argv[0]
+filename_base, _ext = os.path.splitext(self)
+
+# Build a .ipex file in a new temporary dir, each time.
+td = tempfile.mkdtemp()
+hydrated_ipex_file = os.environ.get('PANTS_PTEX_HYDRATE_ONLY_TO_IPEX', None)
+oneshot_run = False
+if not hydrated_ipex_file:
+  oneshot_run = True
+  hydrated_ipex_file = os.path.join(td, '{}.ipex'.format(filename_base))
+
+# Using print(..., file=sys.stderr) complains of invalid syntax for some reason.
+sys.stderr.write('Hydrating {} to {}...\n'.format(self, hydrated_ipex_file))
+
+wrapper_ptex_info = PexInfo.from_pex(self)
+with open_zip(self) as zf:
+  # Populate the pex with the pinned requirements and distribution names & hashes.
+  ipex_info = PexInfo.from_json(zf.read('IPEX-INFO'))
+  ipex_builder = PEXBuilder(pex_info=ipex_info)
+
+  # Populate the pex with the needed code.
+  ptex_info = json.loads(zf.read('PTEX-INFO').decode('utf-8'))
+  for path in ptex_info['code']:
+    zf.extract(path, td)
+    ipex_builder.add_source(os.path.join(td, path), env_filename=path)
```

Comment:
`extract` actually returns the path, so let's use its return value rather than re-implementing it with this `os.path.join` (which is incorrect for a few weird edge cases :))

Reply:
I've made two changes since then:
1. prefix all the paths in the .ptex file with `user_files/` (defined in `_ptex_launcher.py`) -- these are stripped when creating the .ipex.
2. call `zf.extractall()` to extract all the files in one go -- this seems like it might be more performant, but I'm not clear on what edge cases we'd want to check or avoid here from the `os.path.join()` approach?

Reply:
> I've made two changes since then:
> 
> 2. call `zf.extractall()` to extract all the files in one go -- this seems like it might be more performant, but I'm not clear on what edge cases we'd want to check or avoid here from the `os.path.join()` approach?

`extractall` is unsafe because it can unzip _outside_ of the output root: https://docs.python.org/2/library/zipfile.html#zipfile.ZipFile.extractall - we should probably not use it.

Edge-case-wise, `extract` does some normalisation around leading and trailing slashes, drive names, and character-replacement (https://docs.python.org/2/library/zipfile.html#zipfile.ZipFile.extract) which it would be nice to avoid getting bitten by or needing to re-implement :)

Reply:
Thank you for this context!!! My experience points are going through the roof! I will convert to use `.extract()` as you've described!